### ISSUE TITLE
Phase 39.2 — ConceptExplainer: TCAV, Concept Bottleneck Models & Prototype Explanations

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#789]
## Phase 39.2 — ConceptExplainer

> TCAV, Concept Bottleneck Models, Prototype Explanations, Activation Maximization

### Overview

The `ConceptExplainer` provides higher-level, concept-based explanations that go beyond individual features. Implements Kim et al. (2018) TCAV for testing with concept activation vectors, Koh et al. (2020) concept bottleneck models for inherently interpretable predictions, and prototype-based explanations that identify exemplar instances.

### Architecture

```
┌──────────────────────────────────────────┐
│            ConceptExplainer              │
│  ┌──────────┐  ┌──────────┐  ┌────────┐ │
│  │  TCAV    │  │ Concept  │  │Proto-  │ │
│  │  Engine  │  │Bottleneck│  │ type   │ │
│  │          │  │ Builder  │  │ Finder │ │
│  └────┬─────┘  └────┬─────┘  └───┬────┘ │
│       │              │            │      │
│  ┌────▼──────────────▼────────────▼────┐ │
│  │    Concept Activation Cache         │ │
│  │  (layer activations, CAVs, probes)  │ │
│  └────────────────┬────────────────────┘ │
│                   │                      │
│  ┌────────────────▼────────────────────┐ │
│  │    Network Dissection Analyzer      │ │
│  │  (unit-concept alignment scoring)   │ │
│  └─────────────────────────────────────┘ │
└──────────────────────────────────────────┘
```

### Frozen Dataclasses

```python
@dataclass(frozen=True)
class ConceptDefinition:
    """Definition of an interpretable concept."""
    name: str                                # human-readable concept name
    positive_examples: Tuple[int, ...]       # indices of positive examples
    negative_examples: Tuple[int, ...]       # indices of negative examples
    layer_name: str = ""                     # target layer for CAV extraction
    description: str = ""                    # concept description

@dataclass(frozen=True)
class ConceptActivationVector:
    """Trained concept activation vector (CAV)."""
    concept_name: str
    direction: Tuple[float, ...]             # unit vector in activation space
    layer_name: str
    accuracy: float                          # linear classifier accuracy
    statistical_significance: float          # two-sided t-test p-value

@dataclass(frozen=True)
class ConceptExplanation:
    """Concept-level explanation for a prediction."""
    concept_scores: Tuple[Tuple[str, float], ...]  # (concept_name, TCAV_score)
    bottleneck_activations: Tuple[float, ...] = ()  # concept bottleneck values
    nearest_prototypes: Tuple[int, ...] = ()         # prototype instance indices
    prediction: float = 0.0
    fidelity: float = 0.0                            # explanation fidelity score
```

### Protocol Interface

```python
@runtime_checkable
class ConceptExplainerProtocol(Protocol):
    """Protocol for concept-based explanation engines."""

    def train_cav(
        self,
        model: Any,
        concept: ConceptDefinition,
        dataset: np.ndarray,
    ) -> ConceptActivationVector:
        """Train a concept activation vector."""
        ...

    def compute_tcav_score(
        self,
        model: Any,
        cav: ConceptActivationVector,
        instances: np.ndarray,
        target_class: int,
    ) -> float:
        """Compute TCAV score — fraction of inputs with positive conceptual sensitivity."""
        ...

    def explain_with_concepts(
        self,
        model: Any,
        instance: np.ndarray,
        concepts: list[ConceptDefinition],
    ) -> ConceptExplanation:
        """Generate concept-level explanation for a prediction."""
        ...

    def find_prototypes(
        self,
        model: Any,
        dataset: np.ndarray,
        n_prototypes: int = 10,
    ) -> list[int]:
        """Find representative prototype instances."""
        ...
```

### Acceptance Criteria

- [ ] TCAV implementation with linear CAV training and directional derivatives
- [ ] Concept bottleneck model builder with concept → prediction pipeline
- [ ] Prototype selection using MMD-critic or k-medoids in activation space
- [ ] Activation maximization with regularized gradient ascent
- [ ] Network dissection with IoU-based unit-concept alignment
- [ ] Statistical significance testing for TCAV scores (p < 0.05)
- [ ] Concept caching for efficient repeated queries
- [ ] Integration with Phase 28 AttentionFilter for concept-attention mapping

### References

- Kim et al. (2018) — Interpretability Beyond Feature Attribution: TCAV
- Koh et al. (2020) — Concept Bottleneck Models
- Bau et al. (2017) — Network Dissection
- Erhan et al. (2009) — Activation Maximization
- Molnar (2020) — Interpretable Machine Learning, Ch. 10

**Comment by web3guru888** [comment]
## Implementation Notes — ConceptExplainer Code Skeleton

```python
"""asi_build/xai/concept_explainer.py — Phase 39.2"""
from __future__ import annotations

import numpy as np
from dataclasses import dataclass
from typing import Any
from sklearn.linear_model import LogisticRegression
from scipy import stats

@dataclass(frozen=True)
class ConceptDefinition:
    name: str
    positive_examples: tuple[int, ...]
    negative_examples: tuple[int, ...]
    layer_name: str = ""
    description: str = ""

@dataclass(frozen=True)
class ConceptActivationVector:
    concept_name: str
    direction: tuple[float, ...]
    layer_name: str
    accuracy: float
    statistical_significance: float

@dataclass(frozen=True)
class ConceptExplanation:
    concept_scores: tuple[tuple[str, float], ...]
    bottleneck_activations: tuple[float, ...] = ()
    nearest_prototypes: tuple[int, ...] = ()
    prediction: float = 0.0
    fidelity: float = 0.0


class ConceptExplainer:
    """Concept-based explanation engine (Kim et al. 2018, Koh et al. 2020)."""

    def __init__(self, n_random_trials: int = 10, significance_threshold: float = 0.05):
        self._n_random_trials = n_random_trials
        self._significance_threshold = significance_threshold
        self._cav_cache: dict[str, ConceptActivationVector] = {}

    def train_cav(self, model: Any, concept: ConceptDefinition, dataset: np.ndarray) -> ConceptActivationVector:
        """Train a Concept Activation Vector via linear probing."""
        # Extract activations at target layer
        pos_acts = self._extract_activations(model, dataset[list(concept.positive_examples)], concept.layer_name)
        neg_acts = self._extract_activations(model, dataset[list(concept.negative_examples)], concept.layer_name)

        # Train linear classifier
        X = np.vstack([pos_acts, neg_acts])
        y = np.concatenate([np.ones(len(pos_acts)), np.zeros(len(neg_acts))])
        clf = LogisticRegression(max_iter=1000, C=1.0)
        clf.fit(X, y)

        # CAV = normal to decision boundary (unit vector)
        direction = clf.coef_[0]
        direction = direction / np.linalg.norm(direction)

        # Statistical significance via random concept trials
        random_scores = []
        for _ in range(self._n_random_trials):
            perm = np.random.permutation(len(dataset))
            random_pos = perm[:len(concept.positive_examples)]
            random_neg = perm[len(concept.positive_examples):len(concept.positive_examples)+len(concept.negative_examples)]
            rX = np.vstack([self._extract_activations(model, dataset[random_pos], concept.layer_name),
                            self._extract_activations(model, dataset[random_neg], concept.layer_name)])
            ry = np.concatenate([np.ones(len(random_pos)), np.zeros(len(random_neg))])
            rclf = LogisticRegression(max_iter=1000)
            rclf.fit(rX, ry)
            random_scores.append(rclf.score(rX, ry))

        _, p_value = stats.ttest_1samp(random_scores, clf.score(X, y))

        cav = ConceptActivationVector(
            concept_name=concept.name,
            direction=tuple(direction.tolist()),
            layer_name=concept.layer_name,
            accuracy=float(clf.score(X, y)),
            statistical_significance=float(p_value),
        )
        self._cav_cache[concept.name] = cav
        return cav

    def compute_tcav_score(self, model: Any, cav: ConceptActivationVector,
                           instances: np.ndarray, target_class: int) -> float:
        """TCAV score: fraction with positive conceptual sensitivity."""
        activations = self._extract_activations(model, instances, cav.layer_name)
        cav_dir = np.array(cav.direction)
        positive_count = 0
        for i, act in enumerate(activations):
            grad = self._class_gradient(model, instances[i], target_class, cav.layer_name)
            sensitivity = np.dot(grad, cav_dir)
            if sensitivity > 0:
                positive_count += 1
        return positive_count / len(instances)

    def explain_with_concepts(self, model: Any, instance: np.ndarray,
                              concepts: list[ConceptDefinition]) -> ConceptExplanation:
        """Generate concept-level explanation combining TCAV scores."""
        scores = []
        for concept in concepts:
            if concept.name in self._cav_cache:
                cav = self._cav_cache[concept.name]
            else:
                raise ValueError(f"CAV not trained for concept '{concept.name}'. Call train_cav first.")
            tcav = self.compute_tcav_score(model, cav, instance.reshape(1, -1), target_class=0)
            scores.append((concept.name, tcav))

        return ConceptExplanation(
            concept_scores=tuple(scores),
            prediction=float(model.predict(instance.reshape(1, -1))[0]),
        )

    def find_prototypes(self, model: Any, dataset: np.ndarray, n_prototypes: int = 10) -> list[int]:
        """MMD-critic prototype selection (Kim, Khanna & Koyejo 2016)."""
        acts = self._extract_activations(model, dataset, layer_name="penultimate")
        from sklearn.metrics.pairwise import rbf_kernel
        K = rbf_kernel(acts)
        selected = []
        for _ in range(n_prototypes):
            best_idx, best_score = -1, -float("inf")
            for i in range(len(dataset)):
                if i in selected:
                    continue
                score = 2 * K[i].mean() - K[i][selected].mean() if selected else K[i].mean()
                if score > best_score:
                    best_idx, best_score = i, score
            selected.append(best_idx)
        return selected

    def _extract_activations(self, model: Any, data: np.ndarray, layer_name: str) -> np.ndarray:
        """Hook-based activation extraction (placeholder)."""
        # In production: register forward hooks on target layer
        return model.get_activations(data, layer_name)

    def _class_gradient(self, model: Any, instance: np.ndarray, target_class: int, layer_name: str) -> np.ndarray:
        """Gradient of class logit w.r.t. layer activations (placeholder)."""
        return model.get_gradient(instance, target_class, layer_name)
```

### mypy type-check table

| Symbol | Type | Status |
|--------|------|--------|
| `ConceptExplainer.train_cav` | `(Any, ConceptDefinition, ndarray) -> ConceptActivationVector` | ✅ |
| `ConceptExplainer.compute_tcav_score` | `(Any, CAV, ndarray, int) -> float` | ✅ |
| `ConceptExplainer.explain_with_concepts` | `(Any, ndarray, list[ConceptDefinition]) -> ConceptExplanation` | ✅ |
| `ConceptExplainer.find_prototypes` | `(Any, ndarray, int) -> list[int]` | ✅ |
